src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -2,37 +2,47 @@
 
 from pip._internal.req.constructors import install_req_from_line
 from pip._internal.req.req_install import InstallRequirement
+from pip._internal.utils.packaging import get_requires_python
 from pip._internal.utils.typing import MYPY_CHECK_RUNNING
 
 from .base import Candidate
 
 if MYPY_CHECK_RUNNING:
-    from typing import Any, Dict, Optional, Sequence
+    from typing import Any, Callable, Dict, Optional, Sequence
 
     from pip._internal.models.link import Link
     from pip._internal.operations.prepare import RequirementPreparer
     from pip._internal.resolution.base import InstallRequirementProvider
 
+    from pip._vendor.packaging.specifiers import SpecifierSet
     from pip._vendor.packaging.version import _BaseVersion
     from pip._vendor.pkg_resources import Distribution
 
+    from .requirements import RequiredPythonRequirement
+
+    RequiredPythonRequirementProvider = Callable[
+        [SpecifierSet], RequiredPythonRequirement,
+    ]
+
 
 _CANDIDATE_CACHE = {}  # type: Dict[Link, Candidate]
 
 
 def make_candidate(
-    link,             # type: Link
-    preparer,         # type: RequirementPreparer
-    parent,           # type: InstallRequirement
-    make_install_req  # type: InstallRequirementProvider
+    link,              # type: Link
+    preparer,          # type: RequirementPreparer
+    parent,            # type: InstallRequirement
+    make_install_req,  # type: InstallRequirementProvider
+    make_python_req,   # type: RequiredPythonRequirementProvider
```

Comment:
Note for the future - this interface keeps growing. I've added an "extras" argument in the extra implementation PR. We should probably look at refactoring this sooner rather than later. Better not start the new resolver with technical debt already added :-)

Also, we might want to look at decoupling the `requirements` and `candidates` modules a bit better, by exposing functions like this in `base` (or maybe even in `__init__`?)

Reply:
Having a `base` or `common` module sounds like a good idea. I think we’d want to do this before we continue working on more things as well. Maybe try to get the extras implementation in, split out `base` (and have a aggregated object to avoid the argument list keep growing), and come back to this.

Reply:
Resolved with the factory refactor.